# Data Consistency Checks

# Contents
1. Import libraries
2. Import datasets
3. Check imported data
4. Consistency checks

# 1. Import libraries

In [5]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Import datasets

In [7]:
#Define the project path
path = r'/Users/davidgriesel/Documents/0 - CareerFoundry/02 - Data Analytics Immersion/04 - Python Fundamentals for Data Analysts/Instacart Basket Analysis - IC 202409'

## 2.1. Products

In [9]:
#Import dataset
df_products = pd.read_csv(os.path.join(path, '02 - Data', 'Original Data', 'products.csv'), index_col = False)

## 2.2. Orders

In [11]:
#Import dataset
df_orders = pd.read_csv(os.path.join(path, '02 - Data', 'Prepared Data', '04_orders_wrangled.csv'), index_col = 0)

# 3. Check imported data

## 3.1. Products

In [14]:
# Get dimensions
df_products.shape

(49693, 5)

In [15]:
#View first 5 rows of the dataframe
df_products.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

## 3.2. Orders

In [17]:
# Get dimensions
df_orders.shape

(3421083, 6)

In [18]:
#View first 5 rows of the dataframe
df_orders.head()

order_id  user_id  user_order_number  orders_day_of_week  \
0   2539329        1                  1                   2   
1   2398795        1                  2                   3   
2    473747        1                  3                   3   
3   2254736        1                  4                   4   
4    431534        1                  5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

# 4. Consistency checks

## 4.1. Products

### 4.1.1. Accuracy

#### 4.1.1.1. View descriptive statistics

In [23]:
#Review descriptive statistics
df_products.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

##### product_id: There are 49693 records but the max id is 49688 which suggest possible duplicates in the dataset.
##### aisle_id: All products are assigned to aisles. Aisles are numbered from 1 to 134. 
##### department_id: All products are assinged to departments which are numbered 1 to 21.
##### prices: All products are priced. The average price is 9.99 and the median price is 7.10. The standard deviation is very large which suggests extreme outliers in the dataset. Most prices range between 4.10 and 11.20. The maximum price of 99999.00 is suspicious and likely en error. 

#### 4.1.1.2. Find inaccurate values

In [26]:
#Display the frequency of unique values in the prices column
df_products['prices'].value_counts(dropna = True).sort_index(ascending = True)

prices
1.0        116
1.1        296
1.2        297
1.3        286
1.4        321
          ... 
24.8         5
24.9         6
25.0         5
14900.0      1
99999.0      1
Name: count, Length: 242, dtype: int64

##### Identified two extreme outliers (14900.00 and 99999.00)

In [28]:
# Filter and display records where the price is either 14900 or 99999
df_products[df_products['prices'].isin([14900, 99999])]

product_id                      product_name  aisle_id  department_id  \
21554       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
33666       33664             2 % Reduced Fat  Milk        84             16   

        prices  
21554  14900.0  
33666  99999.0

#### 4.1.1.3. Address inaccurate values

In [30]:
#Temporarily set pandas to display all rows
pd.set_option('display.max_rows', None)

In [31]:
#Filter for aisle_id == 108 and product_name contains 'cottage cheese'
df_products[(df_products['aisle_id'] == 108) & (df_products['product_name'].str.contains('cottage cheese', case=False, na=False))]

product_id                                       product_name  \
194           195  Grade A Pasteurized 2% Milkfat Lowfat Cottage ...   
1503         1503                             Low Fat Cottage Cheese   
2943         2943  Low Fat, Small Curd 1% Milkfat No Salt Added C...   
3940         3940                              Nonfat Cottage Cheese   
4991         4991                         Dairy Farms Cottage Cheese   
5015         5015                     Organic Cottage Cheese Classic   
6852         6852         4% Milkfat Cottage Cheese Small Curd 22 OZ   
8285         8285                          Cottage Cheese Large Curd   
8558         8558  Cottage Cheese, Natural Small Curd, 4% Milkfat...   
9808         9808                             Organic Cottage Cheese   
11109       11109      Organic 4% Milk Fat Whole Milk Cottage Cheese   
11203       11203  Cottage Doubles Peach, 2% Milkfat Lowfat Cotta...   
11259       11259                          2% Low-fat Cottage Cheese   
11594       11594              Organic Cottage Cheese Kalamata Olive   
11689       11689                   On the Go Low Fat Cottage Cheese   
13083       13083                   Lowfat Small Curd Cottage Cheese   
15122       15122            Organic Cottage Cheese Sun Dried Tomato   
15285       15285                              Lowfat Cottage Cheese   
16257       16257  2% Milkfat Low Fat Large Curd Pot Style Cottag...   
16703       16703               4% Milkfat Large Curd Cottage Cheese   
19460       19459                          Small Curd Cottage Cheese   
19884       19883                       Small Curd Cottage Cheese 4%   
21274       21273        Grade A Pasteurized Fat Free Cottage Cheese   
21554       21553                   Lowfat 2% Milkfat Cottage Cheese   
22130       22129                   1% Milkfat Lowfat Cottage Cheese   
22731       22730                   Singles 1% Lowfat Cottage Cheese   
25820       25819              Small Curd 4% Milk Fat Cottage Cheese   
26442       26441                            Lf Cottage Cheese 16 Oz   
27712       27710                 Fat Free Small Curd Cottage Cheese   
27792       27790        Small Curd Lowfat 2% Milkfat Cottage Cheese   
28422       28420         Organic Cultured Low Fat 1% Cottage Cheese   
28478       28476                  Organic Small Curd Cottage Cheese   
29128       29126   Organic Strawberry Chia Lowfat 2% Cottage Cheese   
31032       31030               2% Milkfat Large Curd Cottage Cheese   
32499       32497         Lowfat & Strawberry Topping Cottage Cheese   
34232       34230               4% Milkfat Small Curd Cottage Cheese   
35236       35234                          2% Low Fat Cottage Cheese   
35448       35445                          Large Curd Cottage Cheese   
35776       35772                           1% Lowfat Cottage Cheese   
35951       35947   4% Milkfat Small Curd Cottage Cheese with Chives   
36031       36027               Cottage Cheese Large Curd 4% Milkfat   
36343       36339                                     Cottage Cheese   
36365       36361         Organic Cottage Cheese Blueberry Acai Chia   
37564       37560              4% Milkfat Small Curd  Cottage Cheese   
38562       38558                     Organic Low Fat Cottage Cheese   
38612       38608                           Lowfat Cottage Cheese 2%   
38677       38673                           Pineapple Cottage Cheese   
39435       39431                  Cottage Cheese 4% Milkfat Minimum   
40450       40446                          1% Low Fat Cottage Cheese   
41771       41767                               Cottage Cheese 22 Oz   
42311       42307              Organic Reduced Fat 2% Cottage Cheese   
45934       45930                  Small Curd Low Fat Cottage Cheese   
46214       46210          4% Milkfat Min. Small Curd Cottage Cheese   
46330       46326           Grade A Pasturized Lowfat Cottage Cheese   
46791       46787                            Fa

##### In comparison to other simipar products from same aisle, the price is likely to be 14.9

In [33]:
#Filter for aisle_id == 84 and product_name contains 'milk'
df_products[(df_products['aisle_id'] == 84) & (df_products['product_name'].str.contains('Milk', case=False, na=False))]

product_id                                       product_name  \
328           329                       Organic Whole Grassmilk Milk   
877           877                          Ultra-Filtered Whole Milk   
1463         1463                                       Organic Milk   
1729         1729                               2% Lactose Free Milk   
1940         1940                        Organic 2% Reduced Fat Milk   
2090         2090                                         Milk, Goat   
2187         2187                                    Strawberry Milk   
2748         2748                   Organic Reduced Fat Omega-3 Milk   
2873         2873                             Vanilla Macadamia Milk   
2962         2962                      Milk, Reduced Fat, 2% Milkfat   
3583         3583                  Unsweetened Coconut Milk Beverage   
3594         3594                                         Milk Whole   
4210         4210                                         Whole Milk   
4451         4451                 Organic Skim Milk with DHA Omega-3   
4864         4864                    Milk, 2% Reduced Fat, Chocolate   
4879         4879                                     Milk, Fat Free   
5122         5122                Reduced Fat Midnight Chocolate Milk   
5350         5350                           1% Lowfat Chocolate Milk   
5393         5393                       Organic Whole Chocolate Milk   
5514         5514                     Organic Homogenized Whole Milk   
5612         5612                                   Reduced Fat Milk   
5680         5680             All Natural Homogenized Vitamin D Milk   
5785         5785                        Organic Reduced Fat 2% Milk   
6322         6322                             Organic Raw Whole Milk   
6533         6533                             1% Lowfat Organic Milk   
6962         6962                     Organic Whole Milk Plain Kefir   
7329         7329                Almondmilk, Reduced Sugar, Original   
7492         7492                                Cultured Buttermilk   
7791         7791          Organic 1% Milkfat Strawberry Lowfat Milk   
7968         7968                          Select Fat Free Skim Milk   
8346         8346                                     2 Percent Milk   
8522         8522                          Light 3% Milk Fat Egg Nog   
8628         8628                  Unsweetened Flax, Rice, Hemp Milk   
8923         8923              Chocolate Milk Cartons 2% Reduced Fat   
9373         9373                              Organic 2% Buttermilk   
9663         9663                     Strawberry Lowfat Organic Milk   
9749         9749                            Macadamia Milk Original   
10070       10070                            Organic 1% Low Fat Milk   
10208       10208              Probugs Pretty Plain Whole Milk Kefir   
10262       10262                                Enlighten Hemp Milk   
11821       11821                                 Organic Milk Whole   
12122       12122                 Chocolate Flavored Low Fat Milk 1%   
12374       12374                             Belgian Chocolate Milk   
12715       12715                        Organic Vanilla Lowfat Milk   
12968       12968                              Chocolate Cashew Milk   
13166       13166               Organic Milk Reduced Fat, 2% Milkfat   
13225       13225                                        Lowfat Milk   
13249       13249                                          Skim Milk   
13286       13286                           Dairy Pure Fat Free Milk   
13568       13568                         Yogurt Almond Milk Coconut   
13806       13806              1% Milkfat Low Fat Vitamin A & D Milk   
14023       14023                             Milk Chocolate Hobnobs   
14371       14371                                Organic Nonfat Milk   
14882       14882                        2% Milkfat Reduced Fat Milk   
14946       14946  Organic Yummy Veggie Nutriti

##### In comparison to other simipar products from same aisle, the price is likely to be 9.9

In [35]:
# Update the price from 14900 to 14.9
df_products.loc[df_products['prices'] == 14900, 'prices'] = 14.9

# Update the price from 99999 to 9.99
df_products.loc[df_products['prices'] == 99999, 'prices'] = 9.99

In [36]:
# Optionally reset the display option after displaying the DataFrame
pd.reset_option('display.max_rows')#Display the frequency of unique values in the prices column
df_products['prices'].value_counts(dropna = True).sort_index(ascending = True)

prices
1.0     116
1.1     296
1.2     297
1.3     286
1.4     321
       ... 
24.6      4
24.7      5
24.8      5
24.9      6
25.0      5
Name: count, Length: 241, dtype: int64

### 4.1.2. Missing values

#### 4.1.2.1. Find missing values

In [39]:
#Find missing values
df_products.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [40]:
#Extract records where product_name is null and create subset with the results
df_products_nan = df_products[df_products['product_name']. isnull() == True]

In [41]:
#View subset
df_products_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

#### 4.1.2.2. Address missing values

In [43]:
#Extract records where product_name is not null and update the dataframe with the results
df_products_clean = df_products[df_products['product_name']. isnull() == False]

In [44]:
#Get the dimensions of the updated dataframe
df_products_clean.shape

(49677, 5)

In [45]:
#Check results
49693 - 16

49677

### 4.1.3. Mixed-type data

#### 4.1.3.1. Find mixed-type data

In [48]:
#Finding mixed type data
for col in df_products_clean.columns.tolist():
    weird = (df_products_clean[[col]].map(type) != df_products_clean[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_products_clean[weird]) > 0:
        print (col)

In [49]:
#Return the data types of each column in the dataframe
df_products_clean.dtypes

product_id         int64
product_name      object
aisle_id           int64
department_id      int64
prices           float64
dtype: object

#### 4.1.3.2. Address mixed-type data

##### No mixed-type data identified in the dataframe

### 4.1.4. Duplicates

#### 4.1.4.1. Find duplicates

In [54]:
#Find duplicates and create subset with the results
df_products_duplicates = df_products_clean[df_products_clean.duplicated()]

In [55]:
#View subset
df_products_duplicates

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [56]:
#Get the dimensions of the subset
df_products_duplicates.shape

(5, 5)

#### 4.1.4.2. Address duplicates

In [58]:
#Drop duplicates from new dataframe and update the dataframe with results
df_products_clean_no_duplicates = df_products_clean.drop_duplicates()

In [59]:
#Get the dimensions of the updated dataframe
df_products_clean_no_duplicates.shape

(49672, 5)

In [60]:
#Check results
49677 - 5

49672

## 4.2. Orders

### 4.2.1. Accuracy

#### 4.2.1.1. View descriptive statistics

In [64]:
#Review descriptive statistics
df_orders.describe()

order_id       user_id  user_order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06       3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05       1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04       1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00       1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04       5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05       1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05       2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05       1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

##### order_id: Order id's range from 1 to 3421083 with a unique order id assigned sequentially to each record.
##### user_id: Customer id's range from 1 to 206209.
##### order_number: Users placed between 1 and 100 orders with an average of 11 orders per user.
##### orders_day_of_week: Orders are placed throughout the week, but more come in between Sunday and Thursday, with the slightly more being placed earlier in the week.
##### order_hour_of_day: Orders are placed throughout the day, but more come in between 10am and 4pm with most being placed in the early afternoon.
##### days_since_last_order: Users take between 0 and 30 days to place subsequent orders, with most taking about a week or slightly more to palce an order. Count for this column is 206209 less than the others. This is equal to the max for user_id.

### 4.2.2. Missing values

#### 4.2.2.1. Find missing values

In [68]:
#Finding missing values
df_orders.isnull().sum()

order_id                       0
user_id                        0
user_order_number              0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [69]:
#View missing records
df_orders[df_orders['days_since_prior_order'].isnull() == True]

order_id  user_id  user_order_number  orders_day_of_week  \
0         2539329        1                  1                   2   
11        2168274        2                  1                   2   
26        1374495        3                  1                   1   
39        3343014        4                  1                   6   
45        2717275        5                  1                   3   
...           ...      ...                ...                 ...   
3420930    969311   206205                  1                   4   
3420934   3189322   206206                  1                   3   
3421002   2166133   206207                  1                   6   
3421019   2227043   206208                  1                   1   
3421069   3154581   206209                  1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

##### There are 206209 missing records in the 'days_since_last_order' column. View show 'order_number' is 1 for the top and bottom 5 records.

In [71]:
#View all records where order_number is 1 and days_since_last_order is null
df_orders[(df_orders['user_order_number'] == 1) & (df_orders['days_since_prior_order'].isnull() == True)]

order_id  user_id  user_order_number  orders_day_of_week  \
0         2539329        1                  1                   2   
11        2168274        2                  1                   2   
26        1374495        3                  1                   1   
39        3343014        4                  1                   6   
45        2717275        5                  1                   3   
...           ...      ...                ...                 ...   
3420930    969311   206205                  1                   4   
3420934   3189322   206206                  1                   3   
3421002   2166133   206207                  1                   6   
3421019   2227043   206208                  1                   1   
3421069   3154581   206209                  1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

##### Result contains 206209 records where 'order_number' is 1 and 'days_since_last_order' is null. This number agrees to the number of individual user_id's. Can conclude that the variable is null for each user's first order.

#### 4.2.2.2. Address missing values

In [74]:
#Copy the existing dataframe
df_orders_clean = df_orders.copy()

In [75]:
#Create a new column with boolean values
df_orders_clean['first_order'] = df_orders_clean['days_since_prior_order'].isna()

In [76]:
#View all records where order_number is 1 and days_since_last_order is null, and first_order is True
df_orders_clean[(df_orders_clean['user_order_number'] == 1) & (df_orders_clean['days_since_prior_order'].isnull() == True) & (df_orders_clean['first_order'] == True)] 

order_id  user_id  user_order_number  orders_day_of_week  \
0         2539329        1                  1                   2   
11        2168274        2                  1                   2   
26        1374495        3                  1                   1   
39        3343014        4                  1                   6   
45        2717275        5                  1                   3   
...           ...      ...                ...                 ...   
3420930    969311   206205                  1                   4   
3420934   3189322   206206                  1                   3   
3421002   2166133   206207                  1                   6   
3421019   2227043   206208                  1                   1   
3421069   3154581   206209                  1                   3   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
11                      11                     NaN         True  
26                      14                     NaN         True  
39                      11                     NaN         True  
45                      12                     NaN         True  
...                    ...                     ...          ...  
3420930                 12                     NaN         True  
3420934                 18                     NaN         True  
3421002                 19                     NaN         True  
3421019                 15                     NaN         True  
3421069                 11                     NaN         True  

[206209 rows x 7 columns]

In [77]:
#Check frequency of flag column
df_orders_clean['first_order'].value_counts()

first_order
False    3214874
True      206209
Name: count, dtype: int64

In [78]:
#Get dimensions
df_orders_clean.shape

(3421083, 7)

In [79]:
#View first 5 rows of the dataframe
df_orders_clean.head()

order_id  user_id  user_order_number  orders_day_of_week  \
0   2539329        1                  1                   2   
1   2398795        1                  2                   3   
2    473747        1                  3                   3   
3   2254736        1                  4                   4   
4    431534        1                  5                   4   

   order_hour_of_day  days_since_prior_order  first_order  
0                  8                     NaN         True  
1                  7                    15.0        False  
2                 12                    21.0        False  
3                  7                    29.0        False  
4                 15                    28.0        False

##### The null values are logical and deleting the column would result in the loss of valuable information. By creating a flag column, the null values are easily explained while maintaining the integrity of the original dataset

### 4.2.3. Mixed-type data

#### 4.2.3.1. Find mixed-type data

In [83]:
#Finding mixed type data
for col in df_orders.columns.tolist():
    weird = (df_orders[[col]].map(type) != df_orders[[col]].iloc[0].apply(type)).any(axis=1)
    if len(df_orders[weird]) > 0:
        print (col)

In [84]:
#Return the data types of each column in the dataframe
df_orders.dtypes

order_id                    int64
user_id                     int64
user_order_number           int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

#### 4.2.3.2. Address mixed-type data

##### No mixed-type data identified in the dataframe

### 4.2.4. Duplicates

#### 4.2.4.1. Find duplicates

In [89]:
#Find duplicates and create subset with the results
df_duplicate_orders = df_orders_clean[df_orders_clean.duplicated()]

In [90]:
#View results
df_duplicate_orders

Empty DataFrame
Columns: [order_id, user_id, user_order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

#### 4.2.4.2. Address duplicates

##### No duplicates were identified. Any duplicates must always be deleted.

# 5. Export dataframes

## 5.1. Products

In [95]:
#Get the dimensions of the dataframe
df_products_clean_no_duplicates.shape

(49672, 5)

In [96]:
#Export cleaned products dataset
df_products_clean_no_duplicates.to_csv(os.path.join(path, '02 - Data', 'Prepared Data', '05_products_cleaned.csv'))

## 5.2. Orders

In [98]:
#Get the dimensions of the dataframe
df_orders_clean.shape

(3421083, 7)

In [99]:
#Export cleaned orders dataset
df_orders_clean.to_csv(os.path.join(path, '02 - Data', 'Prepared Data', '05_orders_cleaned.csv'))

# Task

## 1. Consistency checks
If you haven’t performed the consistency checks covered in this Exercise on your df_prods dataframe, do so now.

##### Refer: 4.1. Products

## 2. Interpret descriptive statistics
Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further. 
- Tip: Keep an eye on min and max values!

##### Refer: 4.2.1.1. View descriptive statistics

## 3. Find mixed-type data
Check for mixed-type data in your df_ords dataframe.

##### Refer: 4.2.3.1. Find mixed-type data

## 4. Address mixed-type data
If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

##### Refer: 4.2.3.2. Address mixed-type data

## 5. Find missing values
Run a check for missing values in your df_ords dataframe.
- In a markdown cell, report your findings and propose an explanation for any missing values you find.

##### Refer: 4.2.2.1. Find missing values

## 6. Address missing values
Address the missing values using an appropriate method.
- In a markdown cell, explain why you used your method of choice.

##### Refer: 4.2.2.2. Address missing values

## 7. Duplicates
Run a check for duplicate values in your df_ords data.
- In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

##### Refer: 4.2.4.1. Find duplicates

## 8. Address duplicates
Address the duplicates using an appropriate method.
- In a markdown cell, explain why you used your method of choice.

##### Refer: 4.2.4.2. Address duplicates

## 9. Export dataframe
Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

##### Refer: 5. Export dataframes

## 10. Save and submit notebook
Save your Jupyter notebook and submit it here for your tutor to review.

##### Notebook saved and submitted